File: scripts/src/genTlds.ts
import { DOMParser } from "xmldom"
import fs from "fs"
;(async () => {
	let tlds = new Map()

	// get tlds, categories, and managers
	let res = await fetch("https://www.iana.org/domains/root/db")
	let text = await res.text()
	let doc = new DOMParser().parseFromString(text, "text/html")

	let table = doc.getElementsByTagName("table")[0]
	let rows = table.getElementsByTagName("tr")

	for (let i = 1; i < rows.length; i++) {
		let cells = rows[i].getElementsByTagName("td")
		let tld = cells[0].textContent?.trim()
		let category = cells[1].textContent?.trim()
		let manager = cells[2].textContent?.trim().replace(/"/g, "")
		if (tld) {
			tlds.set(tld, { tld, category, manager })
		}
	}

	// get the tlds that has a whois server
	let tldsWithServer = new Map()
	res = await fetch("https://raw.githubusercontent.com/rfc1036/whois/next/tld_serv_list")
	text = await res.text()

	for (let line of text.split("\n")) {
		if (line.length === 0 || line.startsWith("#")) continue
		line = line
			.replace("NONE", "")
			.replace("VERISIGN", "")
			.replace("RECURSIVE", "")
			.replace("ARPA", "")
			.replace("IP6", "")
			.split("#")[0]

		let [tld, server] = line.split(/\s+/, 2)
		if (server && !server.startsWith("WEB") && !server.startsWith("NONE")) {
			tldsWithServer.set(tld, tld)
		}
	}

	res = await fetch("https://raw.githubusercontent.com/rfc1036/whois/next/new_gtlds_list")
	text = await res.text()

	for (let line of text.split("\n")) {
		if (line.length === 0 || line.startsWith("#")) continue

		let tld = "." + line.trim()
		if (tld && !tldsWithServer.has(tld)) {
			tldsWithServer.set(tld, tld)
		}
	}

	// remove tlds that don't have a server
	for (let tld of tlds.keys()) {
		if (!tldsWithServer.has(tld)) {
			tlds.delete(tld)
		}
	}

	// write to file
	const filePath = "../core/src/constants/tlds.rs"
	const linesToKeep = 14

	let existingContent = fs.readFileSync(filePath, "utf8")
	let lines = existingContent.split("\n")
	let preservedLines = lines.slice(0, linesToKeep)
	let newContent = Array.from(tlds.values())
		.map(
			({ tld, category, manager }) =>
				`    "${tld}" => Tld { name: "${tld}", category: "${category}", manager: "${manager}" },`,
		)
		.join("\n")
	let updatedContent = preservedLines.join("\n") + "\n" + newContent + "\n" + ");"
	fs.writeFileSync(filePath, updatedContent)
})()

File: scripts/src/bin/gen_tlds.rs
use reqwest::Client;
use scraper::{Html, Selector};
use std::{error::Error, fs::File, io::Write};

struct Tld {
    name: String,
    category: String,
    manager: String,
}

const IANA_TLDS_URL: &str = "https://www.iana.org/domains/root/db";
const TLDS_FILE: &str = "../apps/server/src/constants/tlds.rs";

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let client = Client::builder().user_agent("curl/7.79.1").build()?;

    let tlds = get_iana_tld_info(&client).await?;
    write_tlds(tlds)?;

    Ok(())
}

async fn get_iana_tld_info(client: &Client) -> Result<Vec<Tld>, Box<dyn Error>> {
    let res = client
        .get(IANA_TLDS_URL)
        .send()
        .await
        .map_err(|e| format!("failed to fetch data from {}: {}", IANA_TLDS_URL, e))?;

    if !res.status().is_success() {
        return Err(format!(
            "failed to fetch data from {} w status code: {}",
            IANA_TLDS_URL,
            res.status()
        )
        .into());
    }

    let text = res.text().await?;
    let doc = Html::parse_document(&text);
    let row_selector = Selector::parse("table#tld-table tbody tr")?;
    let cell_selector = Selector::parse("td")?;

    let mut tlds = Vec::new();
    for row in doc.select(&row_selector) {
        let cells: Vec<_> = row.select(&cell_selector).collect();
        if cells.len() < 3 {
            continue;
        }

        let name = cells[0]
            .text()
            .collect::<String>()
            .trim()
            .to_lowercase()
            .replace(".", "");
        let category = cells[1].text().collect::<String>().trim().to_lowercase();
        let manager = cells[2]
            .text()
            .collect::<String>()
            .trim()
            .to_lowercase()
            .replace("not assigned", "")
            .replace('"', "");

        if name.is_empty() || category.is_empty() || manager.is_empty() {
            continue;
        }

        let tld = Tld {
            name,
            category,
            manager,
        };
        tlds.push(tld);
    }

    if tlds.is_empty() {
        return Err(
            "failed to get iana tlds, mby the website is down or changed structure?".into(),
        );
    }

    Ok(tlds)
}

fn write_tlds(tld_info: Vec<Tld>) -> Result<(), Box<dyn Error>> {
    let mut file = File::create(TLDS_FILE)?;

    writeln!(file, "use phf::{{phf_map, Map}};")?;
    writeln!(file, "use serde::Serialize;")?;
    writeln!(file, "use std::clone::Clone;")?;
    writeln!(file, "use typeshare::typeshare;")?;
    writeln!(file)?;
    writeln!(file, "#[typeshare]")?;
    writeln!(file, "#[derive(Serialize, Clone)]")?;
    writeln!(file, "pub struct Tld {{")?;
    writeln!(file, "    pub name: &'static str,")?;
    writeln!(file, "    pub category: &'static str,")?;
    writeln!(file, "    pub manager: &'static str,")?;
    writeln!(file, "}}")?;
    writeln!(file)?;
    writeln!(file, "pub static TLDS: Map<&str, Tld> = phf_map!(")?;

    for tld in tld_info {
        writeln!(
            file,
            "    \"{}\" => Tld {{ name: \"{}\", category: \"{}\", manager: \"{}\" }},",
            tld.name, tld.name, tld.category, tld.manager
        )?;
    }

    writeln!(file, ");")?;

    Ok(())
}
